### DISCUSSION TITLE
Get base MAC address on ESP32

### DISCUSSION THREAD

**Discussion by liudr** [micropython/micropython#11541] [ESP32] [answered]
I'm attempting to use wiznet w5500 ethernet controller with ESP32. All went well (with a lot of help from here +trial and error) except that I need a MAC address. I assume that lots of ESP32 devices with Ethernet will be on my network for my project so avoiding address collision is crucial. I read from ESP-IDF document that an ESP32 has 4 universally administered MACs, with Ethernet MAC as base MAC + 3:

https://docs.espressif.com/projects/esp-idf/en/release-v4.4/esp32/api-reference/system/system.html#mac-address

So I'd like to know if there is a function that maps the esp_read_mac() call so I can obtain a universally administered address for Ethernet. I know a hack might be to establish a wifi interface and use the .config('mac') to get its MAC address, which should be the base address. Then use that address + 3. If there is a more proper way, without having to set up a wifi station, that would be preferred. Thanks.

**Comment by liudr** [comment]
A bit of info I got:
```
By default, this function takes the eFuse value burned at a pre-defined block (e.g. BLK0 for ESP32, BLK1 for
ESP32-S2) as the base MAC address. Per-interface MAC addresses will be calculated according to the table above.
```
So if there is a read_eFuse() function implemented in ESP32 port, that would work too.
My hack:
```
import network
sta_if=network.WLAN(network.STA_IF)
sta_if.config('mac') # base MAC address
```
One more thing that's bothering me, I can't set the dhcp_hostname on the Ethernet interface! "dhcp_hostname" only exists in WLAN interface but not LAN. Also the network.hostname() only sets that for WLAN. The host name of Ethernet interface is stuck at "espressif". 

I'm trying to establish more stable connections to mission-critical devices with Ethernet. I have lots of ESP32 on the same network identified by their hostnames when they are on WiFi.  So this is not a cosmetic thing for me. Is there any quick way to fix this in code? I know how to compile mp firmware now :D Thanks!

**Reply by glenn20** [reply]
This is not a "proper" way to get the base MAC, but if you want to get the base mac without instantiating a WLAN interface, you can call `import machine; machine.unique_id()`. On ESP32, the unique_id oif the chip is the mac address of the first interface (STA_IF).

**Reply by glenn20** [reply, marked as answer]
Re. the hostname, it appears to not be supported on the micropython LAN driver for esp32. It *might* be as simple as adding:
```c
    esp_exceptions(tcpip_adapter_set_hostname(TCPIP_ADAPTER_IF_ETH, mod_network_hostname));
```
just before the end of `lan_active()` or `get_lan()` in `ports/esp32/network_lan.c`. But, I don't currently have an esp32 device with a lan device to test (I hope to get one very soon).

**Reply by robert-hh** [reply]
That works. It requires as well to add:
`#include "extmod/modnetwork.h"`

**Reply by liudr** [reply]
Thank you @glenn20 and @robert-hh !
I can confirm that the added line works with an adafruit weather ESP32 (wroom). I added the line to both lan_active() and get_lan(). I don't know enough about these two functions to tell why either or both but the project code is very interesting, handling micropython objects in C/C++, which I've only used separately for years with C/C++. Maybe after some hard reading I could start contributing!

**Reply by glenn20** [reply]
Thanks for testing it - glad it worked, and you have discovered the joy of delving into the micropython internals, @liudr  :-). Working out the "best" placement would need a little empirical testing. The ideal location is probably just before the `esp_eth_start()` call in `lan_active()` (so that the hostname is applied just before starting the interface), however, I'm not sure if the interface has to be active first.

**Reply by robert-hh** [reply]
I have added the chaneg to the end of `get_lan()`, so it is set before lan_active() is called. The hostname showed up in the lan router assigned with the IP of that board.
